src/lib/zkapp.ts

```diff
@@ -391,18 +384,23 @@ class SmartContract {
       !methodIntfs.every((m) => m.methodName in ZkappClass._methodMetadata) &&
       !inAnalyze()
     ) {
+      if (snarkContext.get().inRunAndCheck) {
+        let err = new Error(
+          'Can not analyze methods inside Circuit.runAndCheck, because this creates a circuit nested in another circuit'
+        );
+        // EXCEPT if the code that calls this knows that it can first run `analyzeMethods` OUTSIDE runAndCheck and try again
```

Comment:
when does this happen?

Reply:
So, the problem is for smart contracts which use neither `compile` nor `digest`, but just get run in a `Mina.transaction` block. This means that `analyzeMethods` is never called before that block. If `analyzeMethods` is called once, it caches its results so it never has to run the methods again.

Now, the `Reducer.reduce` function actually needs the `analyzeMethods` output, so it calls it under the hood. If it hasn't been run before, it will run inside `Circuit.runAndCheck(f)` inside `createTransaction`: https://github.com/o1-labs/snarkyjs/pull/280/files#diff-1f92503aa572cd2ea7bc843e4e77809f3aef81f49bb25995a80d032127482ffcR106
This fails because of the circuit nesting (AFAIU), so I thought `createTransaction` should do what this comment says -- try to re-run `analyzeMethods` first and then `Circuit.runAndCheck(f)` again.

It would be much simpler to just call `smartContract.analyzeMethods` at the beginning of `createTransaction`. The problem is that `createTransaction` is just executing some function the user gave it, it doesn't know which smart contracts are involved in it. Thats also why we stick a `.bootstrap` function on the error here, so `createTransaction` knows what to call.

I wonder if we can come up with something better eventually :D If we get rid of the `await isReady` requirement, maybe we could just run the equivalent of `analyzeMethods` for one method at a time, right inside the method decorator.